### PR TITLE
DEX-980 [API] Handle incorrect base58 strings in query parameters

### PR DESCRIPTION
**Pull request by ngadiyak**

### CODE REVIEW

File: dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala [hunk 1]
```diff
@@ -19,16 +19,34 @@ object PathMatchers {
...
   }
 
-  val AssetPM: PathMatcher1[Asset] = AkkaMatchers.Segment.flatMap { s =>
-    AssetPair.extractAsset(s).toOption
-  }
-
-  object ByteStrPM extends Base58[ByteStr](ByteStr.decodeBase58(_).toOption)
+  object AssetPM
+      extends Base58[Either[InvalidAsset, Asset]](s =>
+        Option(try Right(AssetPair.extractAsset(s).get)
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala:35]: `AssetPair.extractAsset(s)` returns `Try`. Why do you extract it and catch again? :) Note, there is `toEither` method in `Try`

File: dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala [hunk 2]
```diff
@@ -19,16 +19,34 @@ object PathMatchers {
...
+        catch {
+          case e: Exception =>
+            Left(InvalidAsset(s, e.getMessage))
+        })
+      )
+
+  object OrderPM
+      extends Base58[Either[InvalidBase58String, ByteStr]](s =>
+        Option(try Right(ByteStr.decodeBase58(s).get)
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala:44]: Same

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala (hunk 1 above)
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala]: Fixed

File: dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala (hunk 2 above)
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/routes/PathMatchers.scala]: Fixed

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 3]
```diff
@@ -299,24 +308,28 @@ class MatcherApiRoute(
...
-                val response = rateCache.upsertRate(asset, rate) match {
-                  case None => SimpleResponse(StatusCodes.Created, s"The rate $rate for the asset $assetStr added")
-                  case Some(pv) => SimpleResponse(StatusCodes.OK, s"The rate for the asset $assetStr updated, old value = $pv, new value = $rate")
+    (path(AssetPM) & put & withAuth) { assetOrError =>
+
+      withValidAsset(assetOrError) { a =>
+        entity(as[Double]) { rate =>
+          if (rate <= 0) complete(RateError(error.NonPositiveAssetRate))
+          else
+            withAsset(a) { asset =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:317]: `withValidX` and `withX` is a frequent pattern. We need to incapsulate the logic of `withValidX` inside `withX`

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 4]
```diff
@@ -333,21 +346,26 @@ class MatcherApiRoute(
...
-            case None => RateError(error.RateNotFound(asset), StatusCodes.NotFound)
-            case Some(pv) => SimpleResponse(StatusCodes.OK, s"The rate for the asset $assetStr deleted, old value = $pv")
-          }
-          externalClientDirectoryRef ! WsExternalClientDirectoryActor.Command.BroadcastRatesUpdates(Map(asset -> -1))
-          response
+  def deleteRate: Route = (path(AssetPM) & delete & withAuth) { assetOrError =>
+
+    withValidAsset(assetOrError) {
+      a =>
+        withAsset(a) { asset =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:353]: Here too

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 5]
```diff
@@ -371,10 +389,12 @@ class MatcherApiRoute(
...
     )
   )
-  def getOrderBook: Route = (path(AssetPairPM) & get) { p =>
-    parameters("depth".as[Int].?) { depth =>
-      withAssetPair(p, redirectToInverse = true, depth.fold("")(d => s"?depth=$d")) { pair =>
-        complete(orderBookHttpInfo.getHttpView(pair, MatcherModel.Normalized, depth))
+  def getOrderBook: Route = (path(AssetPairPM) & get) { pairOrError =>
+    withValidAssetPair(pairOrError) { p =>
+      parameters("depth".as[Int].?) { depth =>
+        withAssetPair(p, redirectToInverse = true, depth.fold("")(d => s"?depth=$d")) { pair =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:395]: Too

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 6]
```diff
@@ -393,9 +413,12 @@ class MatcherApiRoute(
       new ApiImplicitParam(name = "priceAsset", value = "Price Asset ID in Pair, or 'WAVES'", dataType = "string", paramType = "path")
     )
   )
-  def getOrderBookStatus: Route = (path(AssetPairPM / "status") & get) { p =>
-    withAssetPair(p, redirectToInverse = true, suffix = "/status") { pair =>
-      complete(orderBookHttpInfo.getMarketStatus(pair))
+  def getOrderBookStatus: Route = (path(AssetPairPM / "status") & get) { pairOrError =>
+    withValidAssetPair(pairOrError) {
+      p =>
+        withAssetPair(p, redirectToInverse = true, suffix = "/status") { pair =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:419]: Too

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 7]
```diff
@@ -562,10 +587,12 @@ class MatcherApiRoute(
       )
     )
   )
-  def cancel: Route = (path(AssetPairPM / "cancel") & post) { p =>
-    withAssetPair(p, formatError = e => OrderCancelRejected(e)) { pair =>
-      unavailableOrderBookBarrier(pair) {
-        handleCancelRequest(Some(pair))
+  def cancel: Route = (path(AssetPairPM / "cancel") & post) { pairOrError =>
+    withValidAssetPair(pairOrError) { p =>
+      withAssetPair(p, formatError = e => OrderCancelRejected(e)) { pair =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:592]: Too

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 8]
```diff
@@ -742,12 +771,14 @@ class MatcherApiRoute(
...
-    (p, publicKeyOrError) =>
-      withValidPublicKey(publicKeyOrError) { publicKey =>
-        withAssetPair(p, redirectToInverse = true, s"/publicKey/$publicKey") { pair =>
-          parameters("activeOnly".as[Boolean].?, "closedOnly".as[Boolean].?) { (activeOnly, closedOnly) =>
-            signedGet(publicKey) {
-              loadOrders(publicKey, Some(pair), getOrderListType(activeOnly, closedOnly, OrderListType.All))
+    (pairOrError, publicKeyOrError) =>
+      withValidAssetPair(pairOrError) { p =>
+        withValidPublicKey(publicKeyOrError) { publicKey =>
+          withAssetPair(p, redirectToInverse = true, s"/publicKey/$publicKey") { pair =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:777]: Too

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 9]
```diff
@@ -930,11 +965,13 @@ class MatcherApiRoute(
...
   )
-  def tradableBalance: Route = (path(AssetPairPM / "tradableBalance" / AddressPM) & get) { (pair, addressOrError) =>
-    withCorrectAddress(addressOrError) { address =>
-      withAssetPair(pair, redirectToInverse = true, s"/tradableBalance/$address") { pair =>
-        complete {
-          askMapAddressActor[AddressActor.Reply.Balance](address, AddressActor.Query.GetTradableBalance(pair.assets))(_.balance.toJson)
+  def tradableBalance: Route = (path(AssetPairPM / "tradableBalance" / AddressPM) & get) { (pairOrError, addressOrError) =>
+    withValidAssetPair(pairOrError) { pair =>
+      withValidAddress(addressOrError) { address =>
+        withAssetPair(pair, redirectToInverse = true, s"/tradableBalance/$address") { pair =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:971]: Too

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 10]
```diff
@@ -990,16 +1027,22 @@ class MatcherApiRoute(
...
-    withAssetPair(p, redirectToInverse = true, s"/$orderId") { _ =>
-      complete {
-        orderDb.get(orderId) match {
-          case Some(order) =>
-            askMapAddressActor[AddressActor.Reply.GetOrderStatus](order.sender, AddressActor.Query.GetOrderStatus(orderId)) { r =>
-              HttpOrderStatus.from(r.x)
+  def orderStatus: Route = (path(AssetPairPM / OrderPM) & get) { (pairOrError, orderIdOrError) =>
+    withValidAssetPair(pairOrError) { p =>
+      withValidOrderId(orderIdOrError) { orderId =>
+        withAssetPair(p, redirectToInverse = true, s"/$orderId") { _ =>
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:1033]: Too

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

**Pull request merged**